### ISSUE TITLE
Design for Global Dashboard

### ISSUE THREAD

**Issue by taoeffect** [okTurtles/group-income#1619]
### Problem

We need a place to:

- Handle DMs between users outside of groups
- Display important news related to Group Income (e.g. feature updates, etc.)
- Possibly host global widgets (for example, possible leaderboards to encourage competition between groups to perform well)

### Solution

Design a global dashboard that has the capability of handling all of this stuff, or at the very least DMs + news.

**Comment by SebinSong** [comment]
**[ Figma link ]** : https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?type=design&node-id=26396-326779&mode=design&t=LHJ3EeI9yZVurepS-0

**Comment by SebinSong** [comment]
Hi @taoeffect , just started researching/working on this issue. One instant question is,

**Q. is this Global dashboard expected to show only when there is a multiple groups for the user? Or do we show this page even when the user has only one group?**

Also, to make the development&testing feel smoother, I'm thinking of splitting this issue into 3 PRs, which are 
1) News&updates -> 2) Direct messages -> 3) Dashboard

which is 'smaller' -> 'larger' order. Does it sound good?

**Comment by taoeffect** [comment]
> Q. is this Global dashboard expected to show only when there is a multiple groups for the user? Or do we show this page even when the user has only one group?

I'm thinking it probably makes sense to always show it.

> which is 'smaller' -> 'larger' order. Does it sound good?

Well, I'm a bit confused because it's not clear to me what the difference between "News & Updates" and "Dashboard" is. Doesn't implementing the "News & Updates" require implementing the scaffolding for the Dashboard as well?

**Comment by SebinSong** [comment]
@taoeffect 
I might have misunderstood the [figma](https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?type=design&node-id=26396-326779&mode=design&t=LHJ3EeI9yZVurepS-0). so are we only caring about this red-box but not the skeleton-like designs above it?

![image](https://github.com/okTurtles/group-income/assets/17641213/a2b3a6e9-c23a-4b1d-921f-03464d63cf42)

**Comment by taoeffect** [comment]
@SebinSong The two designs above it don't have the right names. Sorry that it's not clear! What I meant by the "Global Dashboard" is this entire page, and all of the sub pages within it, so the general layout of that page.

**Comment by SebinSong** [comment]
@taoeffect No worries on that (since there is no designer at the company atm).
Okay, then I can split this task into two PRs which is,

**Step 1)** Entire layout of the Global dashboard (also making sure it looks good in all screen sizes) + adding 'News & Updates' sub-page.
**Step 2)** Adding 'Direct messages' sub-page.

Does it sound good?

**Comment by taoeffect** [comment]
@SebinSong Yep! Perfect 👍

**Comment by SebinSong** [comment]
Hi @taoeffect ,
currently working on the 'step 2' of [here](https://github.com/okTurtles/group-income/issues/1619#issuecomment-2011034723) and need your help with clarifying the scope of work.

I remember you saying in a Slack thread(was not able to find where) that DMing to people outside of the groups of an user is not something we are looking to implement for now. is this right?

So if the above statement is correct, to me it looks like this page will list the existing DMs of a user and clicking the items obviously navigates to the corresponding DM Chat page. is this correct?

Thanks,

**Comment by taoeffect** [comment]
> So if the above statement is correct, to me it looks like this page will list the existing DMs of a user and clicking the items obviously navigates to the corresponding DM Chat page. is this correct?

No, this page should be a global view of all open DMs you have, across all groups, and in the future will include DMs outside of groups too. Clicking the "Direct Messages" should open up a chat view that's very similar to the one we already have, except it shouldn't show any chat channels, only all open DMs you have.

**Comment by SebinSong** [comment]
@taoeffect 

> Clicking the "Direct Messages" should open up a chat view that's very similar to the one we already have

You mean clicking each of these items will lead to showing the chat inside this Global dashboard, instead of navigating to the DM page?(the second screenshot)
(Wouldn't it be also good to just navigate to this DM page instead of working on creating another chat UI that is part of Global-dashboard?)

<img src='https://github.com/okTurtles/group-income/assets/17641213/676f58b7-98fa-419e-8d76-9ffa5624fb2e' width='480'>

[DM page] 
<img src='https://github.com/okTurtles/group-income/assets/17641213/646d1bef-3857-4a71-9bba-63480b36f8d3' width='480'>

**Comment by taoeffect** [comment]
> You mean clicking each of these items will lead to showing the chat inside this Global dashboard, instead of navigating to the DM page?

Correct.

> Wouldn't it be also good to just navigate to this DM page instead of working on creating another chat UI that is part of Global-dashboard

No, because we could have a DM with the same person in two different groups. Which group do you navigate to then?

And again, we will have DMs outside of groups. Which group do you navigate to if there's no group?

**Comment by SebinSong** [comment]
@taoeffect Makes sense. I will work on creating another chatroom UI in the global dashboard then.

**Comment by SebinSong** [comment]
Hi @Silver-IT ,
I'm currently working on implementing this 'Direct Messages' page in the Global dashboard of the app and need your advice/help, as you are someone who has spent more time in the chatroom area than any others.

![image](https://github.com/okTurtles/group-income/assets/17641213/44fb4dd8-da83-4113-b081-972897f22250)

So as you can guess from the screenshot above, obviously the core part of the implementation is to load all the DMs data a user has. I've been spending hours on researching how this is currently happening in `GroupChat.vue` of and this is what I've discovered so far (can be incorrect).

1) `ourDirectMessages` getter holds all the chatroom IDs of a user's open DMs.
2) we need to fetch the chelonia events data using either
- `await sbp('gi.db/archive/load', 'messages/${this.ourIdentityContractId}/${chatroomId}')`
-   `await sbp('chelonia/out/latestHEADInfo', chatroomId)` and `sbp('chelonia/out/eventsBefore', chatroomId, ... `
3) and then we somehow need to process the event chain we got from **2.** above using `sbp('chelonia/in/processMessage', ...)`

etc...

Which is way too complex to re-write and most importantly, currently this logic seems to be written in a way that is closely coupled to the utilities in `ChatMain.vue` (e.g. `this.generateNewChatRoomState()` in there is necessary for **3.** above)

---

In summary,

**Q1.** is there a better way to get all DM data a particular user has? For example, do you think it's possible to create a function like
```
function getAllDMsByUserId (identityContractId) { 
...
  return [
     { chatroomID: ..., messages: [ ...], other attributes },
     ...
  ]
}
```

**Q2.** If the answer for **Q1.** is yes, would you be able to help with that part?
(I think you have the best expertise for these parts so it better be worked on by you if needed.)

**Comment by Silver-IT** [comment]
@SebinSong, it's not efficient to save all the messages in the contract state, and that's why we render the events in `ChatMain.vue` only whenever it's needed. At the moment, we don't have any function that returns the response as you expected because we don't render chatroom contract events unless the user is in Chat page.

Of course, it's possible to make that kind of function, but I need to discuss the solution with @taoeffect beforehand. Will keep you updated soon.

**Comment by SebinSong** [comment]
@Silver-IT Thanks, Alex.

**Comment by SebinSong** [comment]
I have only worked on preparing the UI elements for the 'Direct Messages' sub-page like below (currently rendered with some **dummy data** I created for development). 
Once the solution for the above discussion regarding `loading the DM data from outside of GroupChat.vue` is confirmed & prepared, I will continue to work on this..!

![image](https://github.com/okTurtles/group-income/assets/17641213/93df2c96-8d12-48e2-bd1b-300f69d56cad)

**Comment by SebinSong** [comment]
@Silver-IT So from the mergence of [this PR](https://github.com/okTurtles/group-income/issues/1962) of yours on, it's possible to load DM data from outside the `GroupChat.vue` component, right? can you leave some notes on how to do this?

**Comment by Silver-IT** [comment]
> It's possible to load DM data outside the GroupChat.vue, right? Can you leave some notes on how to do this?

What I did in the PR #1965 is that we added messages in the contract state. So no matter it's DM or not, all the chatroom contracts have the `messages` fields which saves the latest 20 messages. (Exactly, at most 20 messages). I remember the problem was to display the last message.
![image](https://github.com/okTurtles/group-income/assets/57976479/2fe4bbd9-af35-4e88-b1b0-72212b243862)

**Comment by SebinSong** [comment]
Thanks, @Silver-IT

**Comment by taoeffect** [comment]
Note: when implementing this issue, this change from this PR needs to be reverted: https://github.com/okTurtles/group-income/pull/2874/files#diff-a7b7684902a7295ac71c03e9a444304d2ae4aef9d0e5eeb3a9c295800fb65343R56-R61

From `chatroom/getters.js`:

```js
        // The following line ensured that DMs for former members were hidden.
        // Until the DM global dashboard is implemented, this check has been
        // replaced with a check for all members (past and present), so that
        // a conversation doesn't suddenly become inaccessible.
        // // const hasActiveMember = partners.some(memberID => Object.keys(getters.profilesByGroup(groupID)).includes(memberID))
        const hasActiveMember = partners.some(memberID => !!rootState[groupID]?.profiles[memberID])
```

**Comment by SebinSong** [comment]
## Direct Messages in Global Dashboard TODOs

DMs in Global dashboard will likely consists of two parts:
1) The page that lists active DMs
2) Chat Interface for each DM (like `ChatMain.vue`)


### Step 1 - Building UIs that lists active DMs of the user:
This step is about building the page where active DMs are listed like below Figma (Won't necessarily pefectly the Figma. There can be additions to this/reductions from this). The step is mostly about building new UIs/Components so clickable elements won't necessarily lead to actions.

Figma link: https://www.figma.com/design/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=26396-326779&t=EZsSf4QrzTARN1Tb-0

<img width="100%" src="https://github.com/user-attachments/assets/6108b872-1928-4265-abeb-dc2d9b68e779" />


### Step 2 - Building the Chat Interface for each DM.
In this step:

1) Clickable elements built in **Step 1** perform their relevant actions.
2) Will build the chat Interface where user can view/send/receive their DMs. It's highly likely that `ChatMain.vue` reusable component is used in this page too for DRY.

**Comment by SebinSong** [comment]
Hi @taoeffect ,
As you can see from above screenshot, the DM list-item UI element has message preview which Slack does have too.
Here is what I've observed from Slack:

1) All markdown syntaxes are stripped out.
2) If the message spans across multiple lines(has `\n`. regarless of whether contains markdown or not), it only takes the first line and add `...` at the end.
3) If the message is not multi-line, the whole message is taken as the preview text. Overflowing text is truncated and displayed with `...`.

So for example, the preview message of this markdown message:

<img width="480" src="https://github.com/user-attachments/assets/13c9dcb2-d39f-4f3e-bf91-d59817220717" />

is as simple as below, as a result of `1.` and `2.`:

<img width="954" height="198" alt="Image" src="https://github.com/user-attachments/assets/5797e7dd-0aa5-4da2-b8cf-e3fb148a31d9" />

<br>
<br>

This seems to be a good choice from memory usage perspective too, as we take and show only a small portion of the potentially significantly large size text. So would love to implement the preview message in Global DM the same say.
Does it sound good?

Thanks,

**Comment by taoeffect** [comment]
@SebinSong 

> So would love to implement the preview message in Global DM the same say.
Does it sound good?

Yes, it does. 👍 

Sorry I didn't see this reply earlier — (please feel free to ping me on Slack in case I miss one of these messages!)
